core/src/main/scala/com/microsoft/azure/synapse/ml/causal/SyntheticEstimator.scala

```diff
@@ -0,0 +1,234 @@
+      .filter(not(treatment) and not(postTreatment))
+      .select(
+        (outcome -
+          lag(outcome, 1).over(
+            Window.partitionBy(col(getUnitCol)).orderBy(col(getTimeCol))
+          )).as("diff")
+      )
+      .agg(stddev_samp(col("diff")))
+      .head
+
+    val nTreatedPost = data.filter(treatment and postTreatment).count
+    val zeta = math.pow(nTreatedPost, 0.25) * firstDiffStd
+    zeta
+  })
+
+  private[causal] def fitUnitWeights(indexedPreDf: DataFrame,
+                                     zeta: Double,
+                                     fitIntercept: Boolean,
+                                     size: (Long, Long)): (DVector, Double, Seq[Double]) =
+    logVerb("fitUnitWeights", {
+      val outcomePreControl = indexedPreDf.filter(not(treatment))
+        .toDMatrix(TimeIdxCol, UnitIdxCol, getOutcomeCol)
+
+      val outcomePreTreatMean = indexedPreDf.filter(treatment)
+        .groupBy(col(TimeIdxCol).as("i"))
+        .agg(avg(outcome).as("value"))
+        .as[VectorEntry]
+
+      val lambda = if (zeta == 0) 0d else {
+        val t_pre = matrixOps.size(outcomePreControl)._1 // # of time periods pre-treatment
+        zeta * zeta * t_pre
+      }
+
+      solveCLS(outcomePreControl, outcomePreTreatMean, lambda, fitIntercept, size)
+    })
+
+  private[causal] def handleMissingOutcomes(indexed: DataFrame, maxTimeLength: Int): DataFrame = {
+    // "skip", "zero", "impute"
+    getHandleMissingOutcome match {
+      case "skip" =>
+        indexed.withColumn("time_count", count(col(TimeIdxCol)).over(Window.partitionBy(col(UnitIdxCol))))
+          // Only skip units from the control_pre group where there is missing data.
+          .filter(col("time_count") === lit(maxTimeLength) or treatment or postTreatment)
+          .drop("time_count")
+      case "zero" =>
+        indexed
+      case "impute" =>
+        // Only impute the control_pre group.
+        val controlPre = indexed.filter(not(treatment) and not(postTreatment))
+
+        val imputed = imputeTimeSeries(controlPre, maxTimeLength, getOutcomeCol)
+          .withColumn(getTreatmentCol, lit(false))
+          .withColumn(getPostTreatmentCol, lit(false))
+
+        indexed.as("l").join(
+          imputed.as("r"),
+          col(s"l.$UnitIdxCol") === col(s"r.$UnitIdxCol") and col(s"l.$TimeIdxCol") === col(s"r.$TimeIdxCol"),
+          "full_outer"
+        ).select(
+          coalesce(col(s"l.$UnitIdxCol"), col(s"r.$UnitIdxCol")).as(UnitIdxCol),
+          coalesce(col(s"l.$TimeIdxCol"), col(s"r.$TimeIdxCol")).as(TimeIdxCol),
+          coalesce(col(s"l.$getOutcomeCol"), col(s"r.$getOutcomeCol")).as(getOutcomeCol),
+          coalesce(col(s"l.$getTreatmentCol"), col(s"r.$getTreatmentCol")).as(getTreatmentCol),
+          coalesce(col(s"l.$getPostTreatmentCol"), col(s"r.$getPostTreatmentCol")).as(getPostTreatmentCol)
+        )
+    }
+  }
+}
+
+object SyntheticEstimator {
+  val UnitIdxCol = "Unit_idx"
+  val TimeIdxCol = "Time_idx"
```

Comment:
We might want to have a single source of Constants for this work to keep track of em all

Reply:
Let me think more about these two constants. Ideally I want to use either findUnusedColumn or persist them as params.

Reply:
Done. These two constants are replaced with findUnusedColumn.